src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -113,11 +134,331 @@ public void Initialize(ChainedHeader consensusTip)
+        }
+
+        /// <summary>
+        /// Handles situation when block data was considered to be invalid
+        /// for a given header during the partial or full validation.
+        /// </summary>
+        /// <param name="chainedHeader">Chained header which block data failed the validation.</param>
+        /// <returns>List of peer Ids that were claiming chain that contains an invalid block. Such peers should be banned.</returns>
+        public List<int> PartialOrFullValidationFailed(ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            List<int> peersToBan = this.RemoveSubtree(chainedHeader);
+
+            this.RemoveUnclaimedBranch(chainedHeader.Previous);
+
+            this.logger.LogTrace("(-):*.{0}={1}", nameof(peersToBan.Count), peersToBan.Count);
+            return peersToBan;
+        }
+
+        /// <summary>
+        /// Handles situation when consensuses tip was changed.
+        /// </summary>
+        /// <param name="newConsensusTip">The new consensus tip.</param>
+        /// <returns>List of peer Ids that violate max reorg rule.</returns>
+        public List<int> ConsensusTipChanged(ChainedHeader newConsensusTip)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(newConsensusTip), newConsensusTip);
+
+            // Switch consensus tip to the new block header.
+            this.AddOrReplacePeerTip(LocalPeerId, newConsensusTip.HashBlock);
+
+            List<int> peerIdsToResync = this.FindPeersToResync(newConsensusTip);
+
+            // Remove block data for the headers that are too far from the consensus tip.
+            this.CleanOldBlockDataFromMemory(newConsensusTip);
+
+            this.logger.LogTrace("(-)");
+            return peerIdsToResync;
+        }
+
+        /// <summary>Checks each peer's tip if it violates max reorg rule.</summary>
+        /// <param name="consensusTip">Consensus tip.</param>
+        /// <returns>List of peers which tips violate max reorg rule.</returns>
+        private List<int> FindPeersToResync(ChainedHeader consensusTip)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(consensusTip), consensusTip);
+
+            var peerIdsToResync = new List<int>();
+            uint maxReorgLength = this.chainState.MaxReorgLength;
+
+            // Find peers with chains that now violate max reorg.
+            if (maxReorgLength != 0)
+            {
+                foreach (KeyValuePair<int, uint256> peerIdToTipHash in this.peerTipsByPeerId)
+                {
+                    ChainedHeader peerTip = this.chainedHeadersByHash[peerIdToTipHash.Value];
+                    int peerId = peerIdToTipHash.Key;
+
+                    ChainedHeader fork = peerTip.FindFork(consensusTip);
+
+                    // Do nothing in case peer's tip is on our consensus chain.
+                    if ((fork != consensusTip) && (fork != peerTip))
```

Comment:
This condition is not clear can we try to write it better?
Perhaps an en extension method `bool IsOnSameChain(header, out fork)`